### ISSUE TITLE
Still need to unlock key after each screen unlock

### ISSUE THREAD

**Issue by JasonB579** [cruegge/pam-gnupg#5]
I've installed this through the AUR and followed all the directions in the README, as follows:

I added
```
auth     optional  pam_gnupg.so
session  optional  pam_gnupg.so
```
to both `/etc/pam.d/lightdm` and `/etc/pam.d/system-local-login`, as well as just the `auth` line to `/etc/pam.d/i3lock`.

I created `~/.gnupg/gpg-agent.conf` and added `allow-preset-passphrase` to it, along with the line `max-cache-ttl 7200`.

I created `~/.pam-gnupg` and added the keygrips from my two keys (although I really only need this for one anyway).

Each time I unlock my screen, once I do something that needs my gpg key I get another request for it, but then it doesn't seem to time out until the `max-cache-ttl` line, which I believe is the correct timeout. Any help would be greatly appreciated!

Edit: fixed wording mistake

**Comment by cruegge** [comment]
So presetting works on login, but not after unlocking i3lock, do I understand correctly? Do you drop the keys somehow when locking the screen? Unlocking will not work via `auth` when the agent is not already running; only the `session` functions will autostart it, and i3lock does not open a session.

The timeout after `max-cache-ttl` may be caused repeatedly accessing the agent in intervals shorter than `default-cache-ttl`. Each access refreshes the cache timer up to maximum of `max-cache-ttl` iiuc. Otherwise, the keys should expire after `default-cache-ttl` when unlocked via the usual agent password dialog.

**Comment by JasonB579** [comment]
I think you understand correctly? Here's the workflow of exactly what's happening:
1. I log in and get prompted for my GPG password
2. It doesn't request it again until `max-cache-ttl` has passed
3. I lock my screen
4. I unlock it and need to enter my password again.

I tested the `max-cache-ttl` vs `default-cache-ttl` thing with default set to 60 seconds and max set to 120 seconds - it just don't seem to expire after default, even when I don't use the key in those 60 seconds.

In short, I kinda have no clue what's going on...

**Comment by cruegge** [comment]
I've been intending to add debug logging for a long time now, and maybe this is a good opportunity to actually do that. I've added some preliminiary syslog stuff to the `syslog` branch. I didn't push it to `master` since there's no option to turn logging off yet. Do you think you could recompile from that branch and try i3lock again? That should produce some message in your syslog / journal.

**Comment by JasonB579** [comment]
Yep, no problem. Here's the last bit of the journal.
![2019-06-24_19-47](https://user-images.githubusercontent.com/24708457/60059089-119b0900-96b9-11e9-9c39-c7422d37c7ee.png)

**Comment by cruegge** [comment]
Strange, that actually looks like everything is working as intended. Maybe the issue is somewhere else? Does your setup have any non-standard features like a moved gnupg directory (`$GNUPGHOME`)? How do you start the agent and i3lock? Do you clear the agent cache before locking?

Also, the command

~~~
gpg-connect-agent 'keyinfo --list' /bye
~~~

displays information on the keys available to the agent. For me, the output looks like

~~~
S KEYINFO FBDEAD7B0C484CDC85F1CF70352833EB0C921D58 D - - 1 P - - -
~~~

The `1` here means that there's a cached passphrase for this key. What does it show after you unlock the screen?

Regarding the no less strange `max-cache-ttl` stuff: is that connected to `pam-gnupg` at all? What happens if you clear the agent cache (`gpg-connect-agent reloadagent /bye`) and then trigger the passphrase dialog manually, without invoking the pam module?

**Comment by JasonB579** [comment]
I shouldn't have any non-standard features - this is the first time I've done any real messing with gpg or pam besides initially creating the key.

I ordinarily use something called `lock` (https://github.com/ShikherVerma/i3lock-multimonitor) to call i3lock. I looked at how that gets invoked, though, and I didn't think that caused the problem. I also tested this issue by just purely calling `i3lock` and the same issue occurs.

For starting the agent, I'm not entirely sure. I was under the impression that it started automatically when `gpg` or `gpg-connect-agent` was called.

I don't believe I clear the cache before locking, or at least I didn't add anything to do so. Should I?

Running
```
gpg-connect-agent 'keyinfo --list' /bye
```
gives me a similar output. It shows 2 keygrips, one that I added to the `~/.pam-gnupg` file and one that I didn't. The one I added has the same flags as yours.

I'm not sure if the `max-cache-ttl` thing is connected to `pam-gnupg`. I cleared the agent cache and triggered the dialog manually and it seems to now timeout after default. I don't think I changed anything that should do that, but hey I'll take it. That said, I'm still trying to learn about pam, since I hadn't really looked into it before finding this project, and I genuinely don't know how to confirm whether or not the module is running when I try that...

**Comment by cruegge** [comment]
Thanks for the info.

> I was under the impression that it started automatically when gpg or gpg-connect-agent was called.

Yes, that's the case, except that autostart should be disabled in `auth`. I just wanted to check.

> I don't believe I clear the cache before locking, or at least I didn't add anything to do so. Should I?

Not necessarily. Depends on your level of paranoia :) Again, just wanted to check you're not killing the agent or something.

> I genuinely don't know how to confirm whether or not the module is running when I try that.

As long as you don't enter your login passphrase somewhere other than the gpg dialog, it's very likely not running.

Regarding `max-cache-ttl`, I just noticed an interesting detail about the agent behaviour. When you preset the wrong passphrase, the first decryption will fail and you will be prompted for the passphrase again, which will then be cached. However, it will still only timeout after `max-cache-ttl`, i.e. the cache timeout will not be reduced retroactively. That would explain the behaviour you see (except for the fact that everything works fine on login). So just to make sure: do you use the same passphrase for your account and your gpg key?

If you want to make sure the passphrase is correct, you can add

~~~
pam_syslog(pamh, LOG_AUTH | LOG_DEBUG, "token: %s", tok);
~~~

above [this line](https://github.com/cruegge/pam-gnupg/blob/53a830ae295c3e3cc79b6c9e6d6283e72f66f956/src/pam_gnupg.c#L333) in the `syslog` branch (which should probably be inside the `if`, now that I think about it).

**This will dump your passphrase to the syslog**, so make sure you do that on a private machine only.

Alternatively, it might be possible to add

~~~
printf("token: %s\n", tok);
~~~

instead and start i3lock from a terminal to avoid the syslog.

**Comment by JasonB579** [comment]
Jesus, I feel stupid. My account password and GPG passphrase were slightly different. I never thought about it when entering my passwords because I was so used to having them off of each other. I really appreciate all the effort you put into this!

**Comment by cruegge** [comment]
Glad it works for you now :)
